jose/Jwe.ml

```diff
@@ -22,25 +29,98 @@ Steps to create a JWE
 * Base64url encode the Authentication Tag to create the Encoded JWE Authentication Tag.
 * Assemble the final representation: The Compact Serialization of this result is the concatenation of the Encoded JWE Header, the Encoded JWE Encrypted Key, the Encoded JWE Initialization Vector, the Encoded JWE Ciphertext, and the Encoded JWE Authentication Tag in that order, with the five strings being separated by four period ('.') characters. 
 *)
-(*
-  TODO: Implement encryption
+
 let make_cek enc =
   let key_length = Jwa.enc_to_length enc in
   Mirage_crypto_rng.generate (key_length / 8)
   |> Cstruct.to_string
   |> Jwk.make_oct ~use:`Enc
 
-let encrypt ?protected payload ~(jwk : Jwk.priv Jwk.t) =
-  let () = match protected with Some _ -> () | None -> () in
-  let _input_key =
-    match Jwk.get_alg jwk with
-    | `RSA_OAEP -> Ok (Mirage_crypto_rng.generate 32)
-    | _ -> Error `Unsupported_enc
+let pkcs7_pad len cs =
+  let cs_len = Cstruct.len cs in
+  let pad_len = len - cs_len in
+  let padding =
+    String.make pad_len (Cstruct.byte pad_len) |> Cstruct.of_string
   in
-  payload
-*)
-open Utils
-module RSA_OAEP = Mirage_crypto_pk.Rsa.OAEP (Mirage_crypto.Hash.SHA1)
+  Cstruct.append cs padding
+
+let encrypt_payload ~enc ~cek ~init_vector ~aad payload =
+  let iv = Cstruct.of_string init_vector in
+  match enc with
+  | Some `A128CBC_HS256 ->
+      (* RFC 7516 appendix B.1: first 128 bit hmac, last 128 bit aes *)
+      let hmac_key, aes_key = Cstruct.(split (of_string cek) 16) in
+      let key = Mirage_crypto.Cipher_block.AES.CBC.of_secret aes_key in
+      (* B.2 encryption in CBC mode *)
+      Mirage_crypto.Cipher_block.AES.CBC.encrypt ~key ~iv
+        (Cstruct.of_string payload |> pkcs7_pad 512)
```

Comment:
512 comes from where? CBC being a block cipher (with a certain block_size) can only work on blocks of that size. this is the whole reason for padding:
I'd suggest to provide `val pkcs7_pad : Cstruct.t -> int -> Cstruct.t` (as now), but the `int` being the block size:
```OCaml
let pkcs7_pad data block_size =
  let leftover = block_size - (Cstruct.len data mod block_size) in (* this is the number of bytes of data in the last block *)
  let pad_size = block_size - leftover in (* this is the remaining bytes in the last block *)
  let pad = Cstruct.create pad_size in
  Cstruct.memset pad pad_size; (* fills the pad buffer with bytes each containing "pad_size" as value *)
  Cstruct.append data pad
```

The call site (here) is then;
```OCaml
  (pkcs7_pad (Cstruct.of_string payload) Mirage_crypto.Cipher_block.AES.CBC.block_size) (* or directly 16 *)
```

Reply:
Thank you, this is much more robust. In this code `leftover` is actually `pad_size` so I changed it in the actual code.